lib/core/init.js

```diff
@@ -1,57 +1,229 @@
-var _beforeLoadComplete = false;
-
-function RESdoBeforeLoad() {
-	if (document && document.html && document.html.classList) {
-		if (_beforeLoadComplete) return;
-		_beforeLoadComplete = true;
-		// if (beforeLoadDoneOnce) return;
-		// first, go through each module and set all of the options so that if a module needs to check another module's options, they're ready...
-		// console.log('get options start: ' + Date());
-		for (var thisModuleID in modules) {
-			if (typeof modules[thisModuleID] === 'object') {
-
-				// Allow the module to instaniate any dynamic options
-				if (typeof modules[thisModuleID].loadDynamicOptions === 'function') {
-					modules[thisModuleID].loadDynamicOptions();
-				}
-			}
-		}
-		for (var thisModuleID in modules) {
-			if (typeof modules[thisModuleID] === 'object') {
-				RESUtils.options.getOptions(thisModuleID);
-			}
-		}
+RESUtils.init = function() {
+	// Don't fire the script on the iframe. This annoyingly fires this whole thing twice. Yuck.
+	// Also don't fire it on static.reddit or thumbs.reddit, as those are just images.
+	// Also omit blog and code.reddit
+	if (
+			(/\/toolbar\/toolbar\?id/i.test(location.href)) ||
+			(/comscore-iframe/i.test(location.href)) ||
+			(/(?:static|thumbs|blog|code)\.reddit\.com/i.test(location.hostname)) ||
+			(/^[www\.]?(?:i|m)\.reddit\.com/i.test(location.href)) ||
+			(/\.(?:compact|mobile)$/i.test(location.pathname)) ||
+			(/metareddit\.com/i.test(location.href))) {
+		return;
+	}
+	if (sessionStorage.getItem('RES.disabled')) return;
+
+	RESUtils.init._watch();
+	RESUtils.init._loadModules();
+
+	RESUtils.init.await.documentElementReady().done(RESUtils.addBodyClasses);
+	RESUtils.init.await.documentBodyReady().done(RESUtils.addBodyClasses);
+	RESUtils.init.await.beforeLoad().done(RESUtils.addBodyClasses);
+	RESUtils.init.await.go().done(RESUtils.addBodyClasses);
+	RESUtils.init.await.documentBodyReady().done(RESUtils.watchMouseMove);
+	RESUtils.init.await.documentBodyReady().done(RESUtils.init._homePage);
+	RESUtils.init.await.documentBodyReady().done(RESUtils.init._reportVersion);
+	RESUtils.init.await('storage').fail(RESUtils.init._localStorageFailed);
+
+	RESUtils.init.await.go().fail(RESUtils.init._showErrors);
+	RESUtils.init.await.documentBodyReady().always(setTimeout.bind(window, RESUtils.init._showErrors, 10000));
+	RESUtils.deferred.collection([
+		RESUtils.init.await.documentBodyReady(),
+		RESUtils.init.await.go()
+	]).done(RESUtils.initObservers);
+};
+
+RESUtils.init.await = function(type, id) {
+	var def = RESUtils.init.await._deferred(type, id);
+	return def.promise();
+};
+[ 'loadDynamicOptions', 'options', 'beforeLoad', 'documentElementReady', 'documentHeadReady', 'documentBodyReady', 'go', 'afterLoad' ].forEach(function(type) {
```

Comment:
You can probably shorten `documentElementReady`, `documentHeadReady`, and `documentBodyReady` to `documentReady`, `headReady`, and `bodyReady` without confusion.